### PR TITLE
compressor: don't flush on every encodeData() call

### LINKED ISSUE #2909: GZip Filter doesn't work on some clients
**Title**: GZip filter doesn't work on some http clients

**Description**:
Using Gzip filter doens't work some http client (apache http client).

**Repro steps**:

1. Using curl , netcat and gunzip
- Enable GZip Filter:

```
static_resources:
  listeners:
  - address: #http-address
      socket_address:
        address: 0.0.0.0
        port_value: 80
    filter_chains:
    - filters:
      - name: envoy.http_connection_manager
        config:
          codec_type: AUTO
          stat_prefix: ingress_http
          http_filters:
          - name: envoy.gzip
            config:
              memory_level: 9
              compression_level: BEST
          - name: envoy.router
          route_config:
            virtual_hosts: #http-hosts
            - name: local-test-server
              domains:
              - test-server.com
              routes:
              - match:
                  prefix: ""
                route:
                  cluster: local-test-server
  clusters:
  - name: local-test-server
    type: STRICT_DNS
    connect_timeout:
      seconds: 60
      nanos: 0
    lb_policy: ROUND_ROBIN
    hosts:
    - socket_address:
        address: 127.0.0.1
        port_value: 1234
admin:
  access_log_path: /dev/null
  address:
    socket_address:
      address: 0.0.0.0
      port_value: 8001
```

- Run a netcat server -> ./test-server
```
#!/bin/bash

echo -n -e "HTTP/1.1 200 OK\r\nContent-Type: text/plain; charset=UTF-8\r\nContent-Length: 64\r\n\naaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa" | nc -l -p 1234
```

- Curl and try to decompress using gunzip:

```
curl -XPOST http://127.0.0.1:80 -H 'Host: test-server.com' -H 'Accept-Encoding: gzip' | gunzip -
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
gzip: stdin: unexpected end of file
```

```
curl -XGET http://127.0.0.1:80 -H 'Host: test-server.com' -H 'Accept-Encoding: gzip' | gunzip -
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
gzip: stdin: unexpected end of file
```

2. Using apache http client:

- Create a simple http client-server
- Do Request to a server returning payload >30 bytes to trigger gzip compression
- the http client raises exception

Stacktrace of the http-client:

> com.xxx.cham.http.DefaultHttpClient [DEFAULT_HTTP_CLIENT] An error occurred when trying to perform HTTP request. Request: HttpRequest{uri='http://test-api.example.com', headers={}, params={}, method=POST, body='{"example":"tetst"}', contentType=JSON, options=HttpRequestOptions{proxy='null', requestTimeout=-1, ignoreSocketTimeout=false}}::java.io.EOFException: Unexpected end of ZLIB input stream 	at java.util.zip.InflaterInputStream.fill(InflaterInputStream.java:240) 	at java.util.zip.InflaterInputStream.read(InflaterInputStream.java:158) 	at java.util.zip.GZIPInputStream.read(GZIPInputStream.java:117) 	at org.apache.http.client.entity.LazyDecompressingInputStream.read(LazyDecompressingInputStream.java:70) 	at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:284) 	at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:326) 	at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178) 	at java.io.InputStreamReader.read(InputStreamReader.java:184) 	at java.io.BufferedReader.fill(BufferedReader.java:161) 	at java.io.BufferedReader.readLine(BufferedReader.java:324) 	at java.io.BufferedReader.readLine(BufferedReader.java:389) 	at com.airy.cham.http.DefaultHttpClient$HttpSendRunnable.processHttpResponse(DefaultHttpClient.java:90) 	at com.airy.cham.http.DefaultHttpClient$HttpSendRunnable.doPostRequest(DefaultHttpClient.java:154) 	at com.airy.cham.http.DefaultHttpClient$HttpSendRunnable.run(DefaultHttpClient.java:187) 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) 	at java.util.concurrent.FutureTask.run(FutureTask.java:266) 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) 	at java.lang.Thread.run(Thread.java:748)
> --

The GZIP filter's test cases also doesn't seem to include HTTP POST requests

### PR DESCRIPTION
**Pull request by rgs1**
We suspect that the gzip filter is under performing
due to excessive flushes. The changes introduced in #3025 -- which
were an attempt to fix #2909 -- are causing too many flushes
(one per encodeData() call) to ensure the compression ise properly
finished. I think at that point it was possible for encodeData()
to be call with end_stream=false while encodeTrailers() wasn't
handle so Z_FINISH could be missed. This isn't the case anymore,
so there's no need to flush every time we encode a chunk of th
response.

We could make this configurable, but given it's a considerable
perf regression _and_ that we'd still be doing the correct thing
by just calling Z_FINISH when end_stream=true or from encodeTrailers(),
I don't think it's worth it.

This also adds stats for visibility.

Helps with #8448.

Risk Level: medium (relaxes the fix for #2909)
Testing: doing perf tests now and looking at ways to ensure this doesn't cause a regression 

Signed-off-by: Raul Gutierrez Segales <rgs@pinterest.com>

### CODE REVIEW

**Comment by rgs1** [comment]
cc: @dio @gsagula @rojkov

**Comment by rgs1** [comment]
failure seems unrelated (passes for me locally):

```
//test/integration:filter_manager_integration_test                       FAILED in 137.6s
```

File: docs/root/intro/version_history.rst [hunk 1]
```diff
@@ -16,6 +16,7 @@ Version history
   invokes. This effectively makes Envoy act as an egress gateway to AWS Lambda.
 * aws_request_signing: a few fixes so that it works with S3.
 * buffer: force copy when appending small slices to OwnedImpl buffer to avoid fragmentation.
+* compressor: don't flush on every encode data call (helps with issue #8448).e
```
Comment (dio) [inline docs/root/intro/version_history.rst]: Remove `.e`? :)

File: include/envoy/compressor/compressor.h [hunk 2]
```diff
@@ -6,9 +6,9 @@ namespace Envoy {
 namespace Compressor {
 
 /**
- * Compressor state whether to flush the compressor or to finish the compression stream.
+ * Compressor state whether to not flush, flush or to finish the compression stream.
```
Comment (dio) [inline include/envoy/compressor/compressor.h]: Can we have this: flush, no flush and finish?

File: source/extensions/filters/http/common/compressor/compressor.cc [hunk 3]
```diff
@@ -121,7 +121,17 @@ Http::FilterHeadersStatus CompressorFilter::encodeHeaders(Http::ResponseHeaderMa
...
     config_->stats().total_uncompressed_bytes_.add(data.length());
-    compressor_->compress(data, end_stream ? Compressor::State::Finish : Compressor::State::Flush);
+    if (end_stream) {
+      compressor_->compress(data, Compressor::State::Finish);
+      config_->stats().compressed_finished_.inc();
+    } else {
+      // Note: it's fine that we don't flush here, it'll either happen in a future encodeData
+      // call or in encodeTrailers(). This will not cause a regression for
+      // https://github.com/envoyproxy/envoy/pull/3025, given that before encodeTrailers wasn't
+      // handled.
```
Comment (dio) [inline source/extensions/filters/http/common/compressor/compressor.cc]: Awesome. Thanks!

**Comment by dio** [comment]
Sorry @rgs1. Seems like we need to merge master.

**Pull request closed without merging**